protocol/consumer/consumer.go

```diff
@@ -1,21 +1,85 @@
 package consumer
 
-const MaxVersionSupported = 1
+import (
+	"encoding/binary"
+	"errors"
+	"io"
+
+	"github.com/segmentio/kafka-go/protocol"
+)
+
+const MaxVersionSupported = 3
 
 type Subscription struct {
-	Version         int16            `kafka:"min=v0,max=v1"`
-	Topics          []string         `kafka:"min=v0,max=v1"`
-	UserData        []byte           `kafka:"min=v0,max=v1,nullable"`
-	OwnedPartitions []TopicPartition `kafka:"min=v1,max=v1"`
+	Version         int16            `kafka:"min=v0,max=v3"`
+	Topics          []string         `kafka:"min=v0,max=v3"`
+	UserData        []byte           `kafka:"min=v0,max=v3,nullable"`
+	OwnedPartitions []TopicPartition `kafka:"min=v1,max=v3"`
+	GenerationID    int32            `kafka:"min=v2,max=v3"`
+	RackID          string           `kafka:"min=v3,max=v3,nullable"`
+}
+
+type subscriptionBackwardsCompat struct {
+	Version  int16    `kafka:"min=v0,max=v1"`
+	Topics   []string `kafka:"min=v0,max=v1"`
+	UserData []byte   `kafka:"min=v0,max=v1,nullable"`
+}
+
+func (s *subscriptionBackwardsCompat) FromBytes(b []byte) error {
+	// This type is only intended to maintain backwards compatibility with
+	// this library and support other clients in the wild sending
+	// version 1 supscription data without OwnedPartitionsy
+	return protocol.Unmarshal(b, 1, s)
+}
+
+func (s *Subscription) FromBytes(b []byte) error {
+	if len(b) < 2 {
+		return io.ErrUnexpectedEOF
+	}
+	version := readInt16(b[0:2])
+	err := protocol.Unmarshal(b, version, s)
+	if err != nil && version >= 1 && errors.Is(err, io.ErrUnexpectedEOF) {
```

Comment:
Do you think we could either move the comment about backwards capability above here or make these checks in the if statement a separate function like `isBackwardsCompatible`? I had a bit of a hard time understanding it the way the code is structured now